source/manual/architecture-deep-dive.html.md

```diff
@@ -0,0 +1,462 @@
+
+Data is copied from Production to Staging - and from Staging to Integration - every
+24 hours via [automated Jenkins jobs][copy-data-to-staging]. This way our
+environments are always roughly in sync, although it's worth noting that email
+addresses are anonymised and access-limited documents are obfuscated before data is
+copied. The data copying is partly configured in [env-sync-and-backup], however
+we're gradually [configuring this in govuk-puppet][govuk-env-sync] in the move to AWS.
+
+[copy-data-to-staging]: https://deploy.publishing.service.gov.uk/job/Copy_Data_to_Staging/
+[env-sync-and-backup]: https://github.com/alphagov/env-sync-and-backup/
+[govuk-env-sync]: /manual/govuk-env-sync.html
+
+### Deploying
+
+We have detailed docs on [how to deploy an application][deploy]. But what happens
+under the hood?
+
+When a change is merged to an application's 'master' branch, a [Jenkins job]
+runs the tests. If successful, Jenkins pushes a git tag to GitHub (the "release
+tag"), then sends a message to the [govuk-app-deployment] Jenkins job to clone
+the repository, check out the tag and deploy the code to the corresponding nodes
+on Integration using [Capistrano] (a Ruby-based server automation and deployment
+tool). Capistrano does deployments only by default, but can also do deployments
+'with migration' or 'with hard restart', etc, depending on the nature of the change.
+
+A developer must manually trigger a deployment to Staging and Production through
+the [release] app. This uses the same Jenkins/Capistrano pipeline as for
+Integration, but on the Staging and Production Jenkins environments respectively.
+
+Some apps require extra care when deploying. 'static', for example, requires a
+30 minute wait between environment deploys. This is because static is consumed
+directly by GOV.UK applications at runtime - not a gem version like a normal
+dependency - and responses from static are cached for half an hour, so problems
+may not be visible until after this period. There are instructions in the release
+app UI for this.
+
+[Capistrano]: https://capistranorb.com/
+[deploy]: /manual/deploying.html
+[govuk-app-deployment]: https://github.com/alphagov/govuk-app-deployment
+[Jenkins job]: https://github.com/alphagov/govuk-jenkinslib
+[release]: https://github.com/alphagov/release
+
+### Puppet on GOV.UK
+
+![](images/puppet.png)
+
+As discussed in the routing section, we have different 'classes' of machines
+running in the cloud; to recap, the "cache" machines run the "router"
+application. These classes are configured in [govuk-puppet], which uses
+[puppet] under the hood: tooling which configures resources such as files
+and processes.
+
+Puppet runs in a master/agent setup. There is a single "puppetmaster" running
+on its own class of machine, whereas the agents run on all the other machines
+(irrespective of class). Each puppet agent is responsible for running
+[health check] tests against the GOV.UK applications running on the node,
```

Comment:
I don't think Puppet is responsible for running healthchecks. It can however configure Icinga or other services to check things.

Reply:
This is the healthcheck I've linked to: https://github.com/alphagov/govuk-puppet/blob/0b20c7efe7e0c3855d4821e55a914ab577d3b84e/modules/govuk_containers/manifests/app.pp

Perhaps I phrased it poorly. Puppet doesn't define the healthcheck (that's done by the [app itself](https://github.com/alphagov/content-publisher/blob/5b968f1bbfd4fa7e577ea535bb2dc23fcf0c99b8/spec/requests/healthcheck_spec.rb)), but it does appear to be responsible for pinging the healthcheck and triggering an Icinga alert if it fails. How should I rephrase it to be less ambiguous?

Reply:
That `govuk_containers` module isn't a great example, as I'm not sure if we're actually using that Puppet configuration anywhere.

Even so, what's happening there as far as I can see is that Puppet is generating configuration for Icinga to do the checking. The `@@icinga::check` bit relates to this thing [1], I believe the `@@` bit means that the configuration file(s) generated are not ending up on this machine, but somewhere else instead, which in this case will be the monitoring machine that runs Icinga.

1: https://github.com/alphagov/govuk-puppet/blob/master/modules/icinga/manifests/check.pp#L61
2: https://puppet.com/docs/puppet/3.8/lang_exported.html

Reply:
I've done some rewording/clarifying. Happy to swap out the govuk_containers module for a better example if you have one 😄 but hopefully what I have now is a general overview of how it all pieces together.